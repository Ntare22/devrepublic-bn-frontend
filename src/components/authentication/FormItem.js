import React from 'react';
import PropTypes from 'prop-types';
import {
  Form,
} from 'react-bootstrap';


const FormItem = ({
  Col, defaultValue, placeholder, setValue, type,
}) => (
  <Form.Group as={Col}>
    <Form.Control
      required
      type={type}
      defaultValue={defaultValue}
      placeholder={placeholder}
      className="signup-form_field"
      onChange={setValue}
    />
  </Form.Group>
);

FormItem.propTypes = {
  // eslint-disable-next-line react/require-default-props
  Col: PropTypes.elementType,
  defaultValue: PropTypes.string.isRequired,
  placeholder: PropTypes.string.isRequired,
  setValue: PropTypes.func.isRequired,
  type: PropTypes.string.isRequired,
};

export default FormItem;
