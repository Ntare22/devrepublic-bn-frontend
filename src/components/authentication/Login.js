/* eslint-disable react/forbid-prop-types */
/* eslint-disable max-len */
/* eslint-disable consistent-return */
/* eslint-disable react/destructuring-assignment */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import {
  Container, Form, Col, Button, Image, Row,
} from 'react-bootstrap';
import { injectIntl } from 'react-intl';
import PropTypes from 'prop-types';
import { loginAction } from '../../redux/actions/login';
import BackgroundImage from '../../assets/backg.jpg';
import fbImage from '../../assets/f.png';
import gogleImage from '../../assets/g.png';
import ErrorAlert from '../sharedComponents/Alert';
import NavBarComponent from '../sharedComponents/NavbarComponent';
import Footer from '../sharedComponents/Footer';
import '../../scss/login.scss';
import translate from '../languages/Translate';

class Login extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      password: '',
      loading: false,
      invalidEmailFeedback: '',
      emailIsValid: false,
      passwordIsValid: false,
      disableBtn: true,
      visible: false,
      errorMessage: '',
      textColor: '',
      removeHover: 'removeHover',
    };
  }

  handleEmail =(evt) => {
    this.setState({ emailIsValid: false, removeHover: 'removeHover' });
    this.setState({ email: evt.target.value });
    const emailRegex = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    const emailValidation = emailRegex.test(evt.target.value);
    const { passwordIsValid } = this.state;
    if (!emailValidation) {
      return this.setState({ invalidEmailFeedback: translate('Email must be valid'), disableBtn: true });
    }
    this.setState({ invalidEmailFeedback: '' });
    this.setState({ emailIsValid: true });
    if (passwordIsValid) {
      return this.setState({ disableBtn: false, removeHover: '' });
    }
  }

  handlePassword =(evt) => {
    const { emailIsValid, passwordIsValid } = this.state;
    this.setState({ passwordIsValid: false, removeHover: 'removeHover' });
    this.setState({ password: evt.target.value });
    const passwordRegex = /^((?=.*\d)(?=.*[A-Z])(?=.*[a-z])(?=.*\W).{8,30})$/;
    const passValidation = passwordRegex.test(evt.target.value);
    if (!passValidation) {
      return this.setState({ disableBtn: true });
    }
    this.setState({ passwordIsValid: true });
    this.setState({ disableBtn: false });
    if (emailIsValid && passwordIsValid) {
      return this.setState({ disableBtn: false, removeHover: '' });
    }
  }

   handleSubmit = async (event) => {
     event.preventDefault();
     const { email, password } = this.state;
     this.setState({ loading: true, disableBtn: true });
     const actionResposnse = await this.props.loginAction({ email, password });
     if (actionResposnse.payload.status === 401) {
       this.setState({ loading: false });
       return this.setState({ visible: true, errorMessage: 'Wrong email or password. Please Enter valid credential', textColor: 'danger' });
     }
     if (!actionResposnse.payload.isVerified) {
       this.setState({ loading: false });
       this.setState({ email: '', password: '' });
       return this.setState({ visible: true, errorMessage: 'Welcome! Verify your account first.', textColor: 'dark' });
     }
     this.props.history.push('/dashoard');
   }

   render() {
     const {
       loading, emailIsValid, invalidEmailFeedback, disableBtn, errorMessage, visible, textColor, removeHover,
     } = this.state;
     const { intl } = this.props;
     const LOGIN = translate('LOGIN');
     const pleaseWait = translate('Please wait...');
     return (
       <>
         <NavBarComponent />
         <div className="signup-page">
           <Container fluid>
             <Row>
               <Col xs="12" sm="12" md="12" lg="7">
                 <Image className="image" src={BackgroundImage} />
                 <div className="welcome-text">{translate('Welcome to Barefoot Nomad')}</div>
               </Col>
               <Col xs="12" sm="12" md="12" lg="4">
                 <Form validated={false} onSubmit={this.handleSubmit} className="login">
                   <Form.Label>
                     <ErrorAlert message={errorMessage} visible={visible} textColor={textColor} className="validation-error" />
                     <h2 className="login-title">
                       { translate('LOGIN') }
                     </h2>
                   </Form.Label>
                   <Form.Group controlId="validationCustom03" className="login-form">
                     <Form.Group>
                       <div className="text-danger validation-error"><small>{ emailIsValid ? '' : invalidEmailFeedback }</small></div>
                       <Form.Control
                         type="email"
                         value={this.state.email}
                         placeholder="Email"
                         onChange={(e) => this.handleEmail(e)}
                         className="login_field testInput"
                         required
                       />
                     </Form.Group>
                     <Form.Group>
                       <Form.Control
                         className="login_field testInput1"
                         type="password"
                         value={this.state.password}
                         placeholder={intl.formatMessage({ id: 'Password', values: 'Password' })}
                         onChange={(e) => this.handlePassword(e)}
                         required
                       />
                     </Form.Group>
                   </Form.Group>
                   <Button type="submit" className={`${removeHover} login_btn`} disabled={disableBtn}>
                     {
                      loading ? pleaseWait : LOGIN
                    }
                   </Button>
                   <div className="lower-part">
                     <Form.Label>
                       <p className="forget-password">{translate('Forgot password?')}</p>
                       <p className="social-login orUse">{translate('Or Use')}</p>
                       <Image className="social-login logo1" src={gogleImage} />
                       <Image className="social-login" src={fbImage} />
                     </Form.Label>
                   </div>
                 </Form>
               </Col>
             </Row>
           </Container>
         </div>
         <Footer />
       </>
     );
   }
}
const MapStateToProps = ({ user }) => ({
  user,
});
export { Login };
export default injectIntl(connect(MapStateToProps, { loginAction })(Login));

Login.propTypes = {
  loginAction: PropTypes.func.isRequired,
  push: PropTypes.func,
  history: PropTypes.object.isRequired,
  intl: PropTypes.object,
};
