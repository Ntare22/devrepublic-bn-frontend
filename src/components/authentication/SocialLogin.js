import React from 'react';
import PropTypes from 'prop-types';
import googlesvg from '../../assets/google.svg';
import facebookSvg from '../../assets/facebook.svg';
import '../../scss/socialLogin.scss';

const SocialLogin = ({ name }) => (
  <a className="social-login-link" href={`${process.env.BACKEND_LINK}/api/v1/auth/${name}`}>
    <img src={name === 'google' ? googlesvg : facebookSvg} alt={name === 'google' ? 'google' : 'facebook'} />
  </a>
);

SocialLogin.propTypes = {
  name: PropTypes.oneOf(['facebook', 'google']).isRequired,
};
export default SocialLogin;
