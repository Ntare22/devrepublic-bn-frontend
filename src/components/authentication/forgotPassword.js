import  { React, useState } from 'react';

const forgotPasswordComponent = (props) => (
    <div className="container p-5">
        <form action={props.forgottPassword} className="was-validated form-check">
            <div className="form-group">
                <h1 className="text-center"> Forgot your password </h1>
                <input type="email" className="form-control text-lg-center" placeholder="Enter email" required />
            </div>
            <button type="submit" className="btn btn-primary btn-block">Send email</button>
        </form>
    </div>);

export default forgotPasswordComponent;
