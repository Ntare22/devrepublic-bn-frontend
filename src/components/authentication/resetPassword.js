import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { injectIntl } from 'react-intl';
import { connect } from 'react-redux';
import { Container } from 'react-bootstrap';
import Alert from '../sharedComponents/Alert';
import NavBarComponent from '../sharedComponents/NavbarComponent';
import Footer from '../sharedComponents/Footer';
import { resetPasswordAction } from '../../redux/actions/resetPassword';
import '../../scss/password.scss';
import translate from '../languages/Translate';

const resetPassword = (props) => {
  const token = props.location.search.split('token=')[1];
  const [password, setPassword] = useState('');
  const [loading, setLoading] = useState(false);
  const [visible, setVisible] = useState(false);
  const [errorMessage, setErrorMessage] = useState('');
  const [textColor, setTextColor] = useState('');
  const [validationClassName, SetValidation] = useState('form-check text-center');
  const [loadingClassName, setLoadingClassName] = useState('spinner-border-sm');
  const { intl } = props;
  const handleSubmit = async (event) => {
    event.preventDefault();
    setLoading(true);
    setLoadingClassName('spinner-border spinner-border-sm');
    const output = await props.resetPasswordAction(password, token);
    if (output.payload.status === 200) {
      setErrorMessage(translate(output.payload.data.message));
      props.history.push('/dashboard');
    } else {
      setErrorMessage(translate(output.payload.data.error));
      setTextColor('danger');
    }
    setLoading(false);
    setLoadingClassName('spinner-border-sm');
    setVisible(true);
  };
  const handleChange = async (e) => {
    setPassword(e.target.value);
    setVisible(false);
    SetValidation('was-validated form-check text-center');
  };
  return (
    <>
      <NavBarComponent />
      <Container fluid="lg">
        <div className="content">
          <div className="container password-container p-5 justify-content-md-center">
            <form
              className={validationClassName}
              onSubmit={handleSubmit}
            >
              <div className="form-group justify-content-center align-items-center">
                <h1>
                  { translate('Reset your password') }
                </h1>
                <Alert message={errorMessage} visible={visible} textColor={textColor} otherClassName="center col-md-4" />
                <input
                  type="password"
                  className="form-control text-center center col-md-4"
                  pattern="(?=.*\d)(?=.*[A-Z])(?=.*[a-z])(?=.*\W).{8,30}"
                  placeholder={intl.formatMessage({ id: 'New Password', values: 'New Password' })}
                  value={password}
                  onChange={handleChange}
                  required
                />
                <div className="invalid-feedback text-lg-center">
                  { translate('Use a password of at least 8 characters: uppercase, lowercase letters, numbers and special characters') }
                </div>
              </div>
              <button type="submit" className="btn btn-primary col-md-2">
                <span className={loadingClassName} />
            &nbsp;
                {loading ? translate('Loading...') : translate('RESET')}
              </button>
            </form>
          </div>
        </div>
      </Container>
      <Footer />
    </>
  );
};

const MapStateToProps = ({ user }) => ({
  user,
});

export { resetPassword };
export default injectIntl(connect(MapStateToProps, { resetPasswordAction })(resetPassword));

resetPassword.propTypes = {
  resetPasswordAction: PropTypes.func.isRequired,
  location: PropTypes.object.isRequired,
  push: PropTypes.func.isRequired,
  history: PropTypes.object.isRequired,
  intl: PropTypes.object,
};
