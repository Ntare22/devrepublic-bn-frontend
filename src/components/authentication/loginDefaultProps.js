import * as React from 'react';
import { shallow } from 'enzyme';
import { Login as LoginPage } from './Login';

export const defaultProps = {
  history: {},
  user: {},
  push: jest.fn(),
  loginAction: jest.fn(),
  intl: { formatMessage: jest.fn() },
};
export const render = (params, fn = shallow) => {
  const props = { ...defaultProps, ...params };
  return fn(
    <LoginPage {...props} />,
  );
};
