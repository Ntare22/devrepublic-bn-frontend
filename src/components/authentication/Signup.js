/* eslint-disable no-unused-expressions */
/* eslint-disable no-nested-ternary */
/* eslint-disable react/prop-types */
/* eslint-disable no-alert */
/* eslint-disable no-undef */
/* eslint-disable no-console */
import React, { useState } from 'react';
import { connect } from 'react-redux';
import { toast } from 'react-toastify';
import {
  Container, Form, Col, Button, Image, Row,
} from 'react-bootstrap';
import translate from '../languages/Translate';
import { signupAction } from '../../redux/actions/signupAction';
import '../../scss/signup.scss';
import NavBarComponent from '../sharedComponents/NavbarComponent';
import Footer from '../sharedComponents/Footer';
import BackgroundImage from '../../assets/backg.jpg';
import FormItem from './FormItem';
import SocialLogin from './SocialLogin';


export const Signup = (props) => {
  const [firstName, setFirstName] = useState('');
  const [lastName, setLastName] = useState('');
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [isLoggedIn, setLogin] = useState(false);
  const [loading, setloading] = useState(false);
  const [validPassword, setValidPassword] = useState(false);
  const [feedback, setFeedback] = useState('');

  const handleChange = (e) => {
    e.target.placeholder === 'First Name' ? setFirstName(e.target.value)
      : e.target.placeholder === 'Last Name' ? setLastName(e.target.value)
        : e.target.placeholder === 'Email' ? setEmail(e.target.value) : null;
  };

  // eslint-disable-next-line consistent-return
  const handlePassword = (e) => {
    setPassword(e.target.value);

    const passwordRegex = /(?=.*\d)(?=.*[A-Z])(?=.*[a-z])(?=.*\W).{7,30}/;
    const validatePassword = passwordRegex.test(password);

    if (!validatePassword) {
      setValidPassword(true);
      const feedbackMessage = 'Your password must contain atleast 8 character, including a capital letter and a symbol.';
      return setFeedback(feedbackMessage);
    }

    setFeedback('');
    setValidPassword(false);
  };

  const handleSubmit = async (event) => {
    event.preventDefault();

    const data = {
      firstName, lastName, email, password,
    };
    setLogin(!isLoggedIn);
    setloading(true);
    const output = await props.signupAction(data);

    if (output.payload.status === 201) {
      props.history.push('/');
    } else {
      setloading(false);
      const errors = typeof output.payload.error === 'object' ? output.payload.error : [output.payload.error];

      errors.map((e) => toast.error(e, {
        autoClose: 9000,
        position: 'top-right',
      }));
    }
  };

  return (
    <div className="signup-page">
      <NavBarComponent />
      <Container fluid>
        <Row>
          <Col xs="12" sm="12" md="12" lg="8">
            <Image className="image" src={BackgroundImage} />
            <div className="text">Welcome to Barefoot Nomad</div>
          </Col>
          <Col xs="12" sm="12" md="12" lg="4">
            <Form onSubmit={handleSubmit} className="signup-form">
              <Form.Label>
                <h2>
                  {translate('Create an Account')}
                </h2>
              </Form.Label>
              <Form.Row>
                <FormItem
                  Col={Col}
                  type="text"
                  defaultValue={firstName}
                  placeholder="First Name"
                  setValue={handleChange}
                />
                <FormItem
                  Col={Col}
                  type="text"
                  defaultValue={lastName}
                  placeholder="Last Name"
                  setValue={handleChange}
                />
              </Form.Row>
              <Form.Group>
                <FormItem
                  type="email"
                  defaultValue={email}
                  placeholder="Email"
                  setValue={handleChange}
                />
                <FormItem
                  type="password"
                  defaultValue={password}
                  placeholder="Password"
                  setValue={(e) => handlePassword(e)}
                />
                {validPassword ? <div className="feedback-message">{feedback}</div> : null}
              </Form.Group>
              <Button type="submit" className="signup-form_btn">
                {
                  loading ? 'Please wait...' : 'REGISTER'
                }
              </Button>
              <Form.Row className="social-login-signup">
                <div>
                  <p>{translate('Or Use')}</p>
                </div>

                <div className="social-links-signup">
                  <SocialLogin name="google" />
                  <SocialLogin name="facebook" />
                </div>
              </Form.Row>
            </Form>
          </Col>
        </Row>
      </Container>
      <Footer />
    </div>
  );
};

const MapStateToProps = ({ user }) => ({
  user,
});

export default connect(MapStateToProps, { signupAction })(Signup);
