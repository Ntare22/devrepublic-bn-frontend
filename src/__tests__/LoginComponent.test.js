import * as React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import reduxStore from '../store';
import Login from '../components/authentication/Login';
import { defaultProps, render as shallowRender } from '../components/authentication/loginDefaultProps';

describe('Login snapshot', () => {
  const render = (fn = mount) => fn(
    <Provider store={reduxStore}>
      <Login {...defaultProps} />
    </Provider>,
  );
  it('should create a snapshot', () => {
    expect(toJson(render)).toMatchSnapshot();
  });
});


describe('Login functionalities', () => {
  it('should submit a form with all info', async () => {
    const wrapper = shallowRender();
    wrapper.setState({
      email: 'email@wd.defs',
      password: '1278',
      loading: false,
      validateForm: false,
      textColor: '',
      visible: false,
      errorMessage: '',
      invalidEmailFeedback: '',
      invalidPasswordFeedback: '',
      emailIsValid: false,
      removeHover: '',
    });
    const { loginAction } = wrapper.instance().props;
    loginAction.mockReturnValueOnce({ payload: { status: 200, isVerified: true } });
    const form = wrapper.find('.login').first();
    const event = { preventDefault: jest.fn() };
    form.simulate('submit', event);
    wrapper.find('.testInput').first().simulate('change', { target: { value: 'aime@as.dc' } });
    wrapper.find('.testInput1').first().simulate('change', { target: { value: 'aime@as.dc' } });
    wrapper.setProps({ history: { push: jest.fn() } });
  });
  it('renser email and password error messages when credential are not valid', () => {
    const wrapper = shallowRender();
    wrapper.find('.testInput').first().simulate('change', { target: { value: 'aime@as.dc' } });

    const emailRegex = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    wrapper.setState({
      email: 'emailf@wse.wsws,//ws@ws.sx',
      password: '1234',
      invalidEmailFeedback: '',
      invalidPasswordFeedback: '',
      emailIsValid: false,
      passwordIsValid: false,
      disableBtn: true,
      removeHover: 'removeHover',
    });
    const enteredEmail = wrapper.state().email;
    const isEmailValid = emailRegex.test(enteredEmail);
    expect(isEmailValid).toBeFalsy();
    wrapper.setState({
      invalidEmailFeedback: 'Email must be valid',
      disableBtn: true,
    });
    expect(wrapper.state().invalidEmailFeedback).toEqual('Email must be valid');
    wrapper.setState({
      email: 'eaime@sdf.sdc',
      invalidEmailFeedback: '',
    });
    expect(wrapper.state().invalidEmailFeedback).toEqual('');
    const passwordRegex = /(?=.*\d)(?=.*[A-Z])(?=.*[a-z])(?=.*\W).{8,30}/;
    const enteredPassword = wrapper.state().password;
    const isPasswordValid = passwordRegex.test(enteredPassword);
    expect(isPasswordValid).toBeFalsy();
    wrapper.setState({
      invalidPasswordFeedback: 'Minimum 8 character with a special charater, uppcase letter and number',
    });
    expect(wrapper.state().invalidPasswordFeedback).toEqual('Minimum 8 character with a special charater, uppcase letter and number');
    wrapper.setState({
      invalidPasswordFeedback: '',
      emailIsValid: true,
      passwordIsValid: true,
      disableBtn: false,
      email: 'eaime@sdf.sdc',
      invalidEmailFeedback: '',
      password: 'Bien@BAR789',
    });
  });
});
