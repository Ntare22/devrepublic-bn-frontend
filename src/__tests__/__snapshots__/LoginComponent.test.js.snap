// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login snapshot should create a snapshot 1`] = `undefined`;
